### PR TITLE
feat: fiat market data fetching

### PR DESCRIPTION
**Pull request by 0xdef1cafe**
## Description

first of a few PRs to implement multi currency support

this PR doesn't add any user facing changes. it restructures the marketData slice to add fiat
currency support, adds fetching to the AppContext, and updates selectors/tests accordingly.

future PRs will deal with view layer changes/consumption of this data.

- chore: update packages
- chore: add new headers
- feat: add new translations
- feat: add fiat market data
- feat: fetch fiat market data in app context

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

wip

## Risk

anything price related could be wrong

## Testing

a general test of the app, especially charts should be conducted

## Screenshots (if applicable)

devs only - note i've blacklisted market data from the redux tools, see `store.ts` to show it, but this is what the state structure looks like now

image

### CODE REVIEW

File: src/components/AssetHeader/AssetChart.tsx [hunk 1]
```diff
@@ -59,7 +59,7 @@ export const AssetChart = ({ accountId, assetId, isLoaded }: AssetChartProps) =>
   } = useLocaleFormatter({ fiatType: 'USD' })
   const [percentChange, setPercentChange] = useState(0)
   const alertIconColor = useColorModeValue('blue.500', 'blue.200')
-  const [timeframe, setTimeframe] = useState(HistoryTimeframe.MONTH)
+  const [timeframe, setTimeframe] = useState(DEFAULT_HISTORY_TIMEFRAME)
```
Comment (0xdef1cafe) [inline src/components/AssetHeader/AssetChart.tsx:62]: use a centralized source of truth for the default allows us to prefetch this data

File: src/context/AppProvider/AppContext.tsx [hunk 2]
```diff
@@ -300,34 +304,33 @@ export const AppProvider = ({ children }: { children: React.ReactNode }) => {
...
-
-    const fetchMarketData = () => {
+    const fetchMarketData = () =>
       portfolioAssetIds.forEach(assetId => {
-        dispatch(marketApi.endpoints.findByAssetId.initiate(assetId, { forceRefetch: true }))
+        dispatch(marketApi.endpoints.findByAssetId.initiate(assetId))
+        const timeframe = DEFAULT_HISTORY_TIMEFRAME
+        const payload = { assetId, timeframe }
+        dispatch(marketApi.endpoints.findPriceHistoryByAssetId.initiate(payload))
       })
```
Comment (0xdef1cafe) [inline src/context/AppProvider/AppContext.tsx:315]: simplified this logic, it was a mess, and does the same thing

File: src/context/AppProvider/AppContext.tsx [hunk 3]
```diff
@@ -300,34 +304,33 @@ export const AppProvider = ({ children }: { children: React.ReactNode }) => {
...
 
-    // clear the old timer
-    if (marketDataIntervalId) {
-      clearInterval(marketDataIntervalId)
-      setMarketDataIntervalId(undefined)
-    }
+    fetchMarketData() // fetch every time assetIds change
+    // refetch every two minutes
+    return () => clearInterval(setInterval(fetchMarketData, 1000 * 60 * 2))
+  }, [portfolioAssetIds, dispatch])
```
Comment (0xdef1cafe) [inline src/context/AppProvider/AppContext.tsx:320]: simplified this existing logic, it was a mess, and does the same thing but doesn't overfetch now

File: src/context/AppProvider/AppContext.tsx [hunk 4]
```diff
@@ -300,34 +304,33 @@ export const AppProvider = ({ children }: { children: React.ReactNode }) => {
...
 
-    const MARKET_DATA_REFRESH_INTERVAL = 1000 * 60 * 2 // two minutes
-    setMarketDataIntervalId(setInterval(fetchMarketData, MARKET_DATA_REFRESH_INTERVAL))
-
-    // marketDataIntervalId causes infinite loop
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [portfolioAssetIds, setMarketDataIntervalId, dispatch])
+  /**
+   * fetch forex spot and history for user's selected currency
```
Comment (0xdef1cafe) [inline src/context/AppProvider/AppContext.tsx:323]: new fiat market data fetching

File: src/hooks/useBalanceChartData/useBalanceChartData.ts [hunk 5]
```diff
@@ -176,9 +161,7 @@ export const bucketEvents = (
 type FiatBalanceAtBucketArgs = {
   bucket: Bucket
   portfolioAssets: PortfolioAssets
-  priceHistoryData: {
-    [k: AssetId]: HistoryData[]
-  }
+  priceHistoryData: PriceHistoryData
```
Comment (0xdef1cafe) [inline src/hooks/useBalanceChartData/useBalanceChartData.ts:164]: use the type def from the origin of the data

File: src/hooks/useBalanceChartData/useBalanceChartData.ts [hunk 6]
```diff
@@ -396,7 +381,8 @@ export const useBalanceChartData: UseBalanceChartData = args => {
     // data prep
     const noDeviceId = isNil(walletInfo?.deviceId)
     const noAssetIds = !assetIds.length
-    if (noDeviceId || noAssetIds || priceHistoryDataLoading) {
+    const noPriceHistoryData = isEmpty(priceHistoryData)
+    if (noDeviceId || noAssetIds || priceHistoryDataLoading || noPriceHistoryData) {
```
Comment (0xdef1cafe) [inline src/hooks/useBalanceChartData/useBalanceChartData.ts:385]: this was obscure and previously wrong, we'd try and compute without price history data, type def update caught this

File: src/lib/charts.ts [hunk 7]
```diff
@@ -1,19 +1,34 @@
...
+export const priceAtDate: PriceAtDate = ({ date, priceHistoryData }): number => {
+  const { length } = priceHistoryData
+  if (!length) {
+    moduleLogger.warn('priceAtDate: no price history data - something will be wrong!')
+    return 0
+  }
+  // https://lodash.com/docs/4.17.15#sortedIndexBy - binary search (O(log n)) rather than O(n)
+  const i = sortedIndexBy(priceHistoryData, { date, price: 0 }, ({ date }) => Number(date))
+  if (i >= length) return priceHistoryData[length - 1].price
+  return priceHistoryData[i].price
```
Comment (0xdef1cafe) [inline src/lib/charts.ts:33]: same function as before from balance chart data - now required in selectors too, simplified and safer

File: src/state/slices/preferencesSlice/preferencesSlice.ts [hunk 8]
```diff
@@ -27,6 +29,7 @@ const initialState: Preferences = {
   },
   selectedLocale: simpleLocale(),
   balanceThreshold: '0',
+  selectedCurrency: 'USD',
```
Comment (0xdef1cafe) [inline src/state/slices/preferencesSlice/preferencesSlice.ts:32]: required for language selection, not yet used in view layer, needed to bring over selectors from monster branch without rewriting

File: src/state/store.ts [hunk 9]
```diff
@@ -47,13 +47,22 @@ export const clearState = () => {
 
 /**
  * These actions make the redux devtools crash. Blacklist them from the developer tools.
+ * remove the blacklist for local debugging, but don't commit it
```
Comment (0xdef1cafe) [inline src/state/store.ts:50]: market data is getting obese and bogging down dev tools - this makes it snappy for most devving, and can be disabled by commenting bits out as required

File: src/state/slices/marketDataSlice/marketDataSlice.tsx [hunk 10]
```diff
@@ -31,30 +50,50 @@ const initialPriceHistory: PriceHistoryByTimeframe = {
...
+  crypto: {
+    byId: {},
+    ids: [],
+    priceHistory: initialPriceHistory,
+  },
+  fiat: {
+    byId: {},
+    ids: [],
+    priceHistory: initialPriceHistory,
+  },
```
Comment (0xdef1cafe) [inline src/state/slices/marketDataSlice/marketDataSlice.tsx:62]: this is the crux of the PR - we're adding fiat price history to the market data state, and namespacing under `crypto` and `fiat` respectively. everything else is a result of this.

File: src/context/AppProvider/AppContext.tsx (hunk 3 above)
Comment (elmutt) [reply src/context/AppProvider/AppContext.tsx:320]: Will this correctly clear the interval since its not in state anymore?

its clearing the new interval?

**Comment by elmutt** [comment]
Is it possible for the chart changes happen in another PR? might make this one easier to review if they are able to happen separately

**Comment by 0xdef1cafe** [comment]
> 

not really - they're bug fixes as a result of more accurate type defs

File: src/hooks/useBalanceChartData/useBalanceChartData.ts [hunk 11]
```diff
@@ -44,22 +45,6 @@ import { useAppSelector } from 'state/store'
 
 import { includeStakedBalance, includeTransaction } from './cosmosUtils'
 
-type PriceAtBlockTimeArgs = {
```
Comment (0xdef1cafe) [inline src/hooks/useBalanceChartData/useBalanceChartData.ts:47]: this now lives in lib/charts

File: src/hooks/useBalanceChartData/useBalanceChartData.ts [hunk 12]
```diff
@@ -195,7 +178,7 @@ const fiatBalanceAtBucket: FiatBalanceAtBucket = ({
   return Object.entries(crypto).reduce((acc, [assetId, assetCryptoBalance]) => {
     const assetPriceHistoryData = priceHistoryData[assetId]
     if (!assetPriceHistoryData?.length) return acc
-    const price = priceAtBlockTime({ assetPriceHistoryData, date })
+    const price = priceAtDate({ priceHistoryData: assetPriceHistoryData, date })
```
Comment (0xdef1cafe) [inline src/hooks/useBalanceChartData/useBalanceChartData.ts:181]: same function, renamed and imported

File: src/context/AppProvider/AppContext.tsx (hunk 3 above)
Comment (0xdef1cafe) [reply src/context/AppProvider/AppContext.tsx:320]: as discussed this is correct

Comment (gomesalexandre) [reply src/context/AppProvider/AppContext.tsx:320]: Tested it out, this is still working as the previous implementation

Comment (gomesalexandre) [reply src/context/AppProvider/AppContext.tsx:320]: Actually I stand corrected, the initial fetch is still working as intended, but I tried testing against develop many times, and the subsequent, 2mn interval isn't working anymore it seems (tried logging in `fetchMarketData` body)

**Review by gomesalexandre — COMMENTED** [review]
Did a full regression testing against prod on:

- Asset Market/Balance charts/data
- Tx history fiat value
- Swapper fiat rate
- Portfolio balance charts
- Portfolio accounts fiat value
- DeFi Overview/Earn fiat values (TVL, balances)

Didn't seem to find any regression except for the 2mn marketData interval that's working on dev, see comment

File: src/context/AppProvider/AppContext.tsx (hunk 3 above)
Comment (0xdef1cafe) [reply src/context/AppProvider/AppContext.tsx]: derp, of course, the intervel was not being set until the cleanup function ran, fixed.

**Comment by 0xdef1cafe** [comment]
> Did a full regression testing against prod on:
> 
> * Asset Market/Balance charts/data
> * Tx history fiat value
> * Swapper fiat rate
> * Portfolio balance charts
> * Portfolio accounts fiat value
> * DeFi Overview/Earn fiat values (TVL, balances)
> 
> Didn't seem to find any regression except for the 2mn marketData interval that's working on dev, see comment

thank you bigly gomie @gomesalexandre, fixed the market data refetch

**Review by gomesalexandre — DISMISSED** [review]
LGTM, tested again against prod, no regressions as noted previously and the interval is now properly set with market data re-fetched every 2mn 🍬

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
Looking good, mainly non-blocking requests.

Potentially blocking - when clicking on an asset, this line logs the warning several hundred times and rugs my console:

https://github.com/shapeshift/web/blob/39e429f01955f8bc7d8b37ff0666b1eebf533f8d/src/lib/charts.ts#L27

**Comment by elmutt** [comment]
Tested locally, no regression noticed. Looks good!

**Review by elmutt — APPROVED** [review]

**Review by cjthompson — APPROVED** [review]

**Pull request merged**